trinity/config.py

```diff
@@ -440,14 +448,49 @@ def nodekey(self, value: Union[bytes, PrivateKey]) -> None:
                 f"`PrivateKey` instance: got {type(self._nodekey)}"
             )
 
+    @contextmanager
+    def process_id_file(self, process_name: str):  # type: ignore
+        with PidFile(process_name, self.data_dir):
+            yield
+
     @classmethod
-    def from_parser_args(cls, parser_args: argparse.Namespace) -> 'TrinityConfig':
+    def from_parser_args(cls,
+                         parser_args: argparse.Namespace,
+                         sub_config_types: Iterable[Type['BaseSubConfig']]) -> 'TrinityConfig':
         """
         Helper function for initializing from the namespace object produced by
         an ``argparse.ArgumentParser``
         """
         constructor_kwargs = construct_trinity_config_params(parser_args)
-        return cls(**constructor_kwargs)
+        trinity_config = cls(**constructor_kwargs)
+
+        trinity_config.initialize_sub_configs(parser_args, sub_config_types)
+
+        return trinity_config
+
+    def initialize_sub_configs(self,
```

Comment:
I think I'm :-1: on this API (for now while we work out the shape of this).  What do you think about the following.

```python
trinity_config = cls(**constructor_kwargs)

for app_config_type in app_config_types:
    app_config = app_config_type.from_parser_args(parser_args, trinity_config)
    trinity_config.add_app_config(app_config)
```

The code above also implicitely suggests that the `AppConfig` classes may each have their own unique `__init__` implementations but that they each *know* how to construct themselves from `parser_args`.  I like this distinction because it removes tight coupling between the `AppConfig` and the `parser_args` with-respect-to instantiation which I think is a better API to support testing.

Thoughts?

Reply:
I like that! :heart: :+1:

Reply:
The only downside of this though is that then, it is up to the `BaseSubConfig` what to do with the passed `trinity_config`. That means, there is no defined way how to access the `base_config` from the sub config.

E.g.

```python
class TrinityEth1SubConfig(BaseSubConfig):

    @property
    def node_class(self) -> Type['Node']:
        from trinity.nodes.full import FullNode
        from trinity.nodes.light import LightNode

        if self.base_config.is_full_mode:                   # <--- NEEDED HERE
            return FullNode
        elif self.base_config.is_light_mode:
            return LightNode
        else:
            raise NotImplementedError("Only full and light sync modes are supported")
```

Unless the `BaseSubConfig` dictates these as properties but even then it is more like a *suggestion* rather than a guarantee that the base config will be accessible at `base_config`.

Reply:
WRT to how an `AppConfig` gets access to the `TrinityConfig`, I think that's something up to the `AppConfig` to determine and I don't think it needs to be standardized.  Also, in your example above, I would expect the `is_full_mode` and `is_light_mode` to live in the `Eth1AppConfig` class.

Reply:
Ah, I just made a comment before reading this one.

>I think that's something up to the AppConfig to determine and I don't think it needs to be standardized

Mmh...ok. I think I would probably favor the initial PR design where we do end up with a standardized access to `base_config` and `args` in each `AppConfig` and rather ditch the "let them be free with the `__init__` method idea but I don't feel strongly about it.

It just strikes me as something annoying to be forced to rely on conventions with this, when we could rely on the code itself. Your call.